### ISSUE TITLE
eliminate additional reader page in URL for list/reader items

### ISSUE THREAD

**Issue by frontendschlampe** [contao/contao#220]
Is it possible to eliminate the additional reader page in URL for News, Events, FAQ etc. Here an example:

**Current behavior:**

Listing: ```https://demo.contao.org/en/news.html```
Details: ```https://demo.contao.org/en/news-detail/new-contao-versions-in-short-intervalls.html```

**Correct behavior:**

Listing: ```https://demo.contao.org/en/news.html```
Details: ```https://demo.contao.org/en/news/new-contao-versions-in-short-intervalls.html```

If I want to open the URL ```https://demo.contao.org/en/news-detail.html``` I will get a 404 error, which is correct, because the ```news-detail.html``` is just a virtual page.

I know, that I can put the list and reader module on the same page, but if list and reader page have various content besides list/reader module, you have to create two different pages. With the PR https://github.com/contao/contao/pull/186 we will fix the problem in breadcrumb navigation. Now we have to fix it in URL, too.

On contao.org you also see the problem: on the page https://contao.org/en/news.html you will find the news list with headline h1 "News about Contao" and intro text "Read the official Contao announcements." On the detail page https://contao.org/en/news/contao-manager-1_1_0.html you also find headline h1 and intro text. But it shouldn't be there. The headline of the news "Contao Manager 1.1.0 is available" should be the h1. The headline and intro text is useless for the detail page.

Any idea?

/cc @dmolineus

**Comment by fritzmg** [comment]
> **Current behavior:**
> 
> Listing: `https://demo.contao.org/en/news.html`
> Details: `https://demo.contao.org/en/news-detail/new-contao-versions-in-short-intervalls.html`
> 
> **Correct behavior:**
> 
> Listing: `https://demo.contao.org/en/news.html`
> Details: `https://demo.contao.org/en/news/new-contao-versions-in-short-intervalls.html`

Just use the following on your `news.html` page:
```php
<?php

if (\Contao\Input::get('auto_item')) {
    echo \Contao\Controller::getFrontendModule(…); // the reader module
}
else {
    echo \Contao\Controller::getFrontendModule(…); // the list module
}
```

**Comment by leofeyer** [comment]
Isn't this what we have implemented in #161?

**Comment by fritzmg** [comment]
I think you mean https://github.com/contao/news-bundle/pull/39 :)

**Comment by leofeyer** [comment]
> On contao.org you also see the problem: on the page https://contao.org/en/news.html you will find the news list with headline h1 "News about Contao" and intro text "Read the official Contao announcements." On the detail page https://contao.org/en/news/contao-manager-1_1_0.html you also find headline h1 and intro text. But it shouldn't be there. The headline of the news "Contao Manager 1.1.0 is available" should be the h1. The headline and intro text is useless for the detail page.

#161 is about overwriting the page title. If you used `{{page::pageTitle}}` in the H1 of the page, it would be replaced with the news title now.

**Comment by frontendschlampe** [comment]
@fritzmg I don't want to add list and reader on the same page.
@leofeyer I don't mean the page title. I mean the headline and intro text in the content area.

I have two pages: one news list page and one news detail page, because the information on both page should be different. For example:

The news list page includes in main area an intro headline, a text and the news list module for all news archives. In right column is a news archive navigation and a contact teaser.

The detail page don't have any intro headline and text in main area, but the news detail module. In right column there's no news archive navigation, but a latest news list for the same news archive of the news in details and a contact teaser of the editor of the news.

In the most cases the requirements above are wanted, that's why I can't place list and reader on the same page!

**Comment by fritzmg** [comment]
Currently this can only be done with an (~ugly) workaround. Isotope does the same, when using a different page for the product reader: the generated URL for the detail page will still look like it's the auto_item for the list page - however via the `getPageIdFromUrl` hook, Isotope tells Contao to actually render a different page.

I am not sure how a more general solution for Contao would look like.

**Comment by frontendschlampe** [comment]
@fritzmg yes ... I know. That's why I create this issue! ;-)

**Comment by leofeyer** [comment]
Again, just use `{{page::pageTitle}}` and `{{page::pageDescription}}` as headline and text on the reader page. It will be replaced by the news title and description as of Contao 4.7.

**Comment by frontendschlampe** [comment]
@leofeyer Again ... title and description are irrelevant ... please check my comment https://github.com/contao/contao/issues/220#issuecomment-445206152

btw. I don't want to have my meta information in content of my page ...

**Comment by fritzmg** [comment]
@leofeyer this is about something more fundamental which is currently not easily possible in Contao. @frontendschlampe wants to have two completely different page layouts for the list view and the reader view - yet the URL (from Contao's perspective) should stay the same. As I said, the Isotope shop from @terminal42 currently works around this by hooking into `getPageIdFromUrl` and then just let Contao render a different page under certain circumstances. The circumstances being:

* The URL has an auto_item fragment for a product.
* The first fragment of the URL is the page with the product list.
* Said page has a "reader page" configured in its settings.

Via the `getPageIdFromUrl` hook the ID/alias of the "reader page" is then simply returned as the first fragment, instead of the original fragment. Thus you can have

* `https://example.org/shop.html` with a product list module,
* `https://example.org/shop/product-lorem-ipsum.html` which is actually not the original `shop` page, but a different page altogether, with a different layout, containing the product reader.

**Comment by dmolineus** [comment]
Since Contao 4.6 there is the `requireItem` in a page (https://github.com/contao/contao/pull/186#issuecomment-445185083). Maybe it could be used to adopt the page id determination so that the detail page is rendered even it's not present in the url - the same as isotope does according to @fritzmg.

But probably there would some edge cases which might lead to unexpected behaviour. For instance, only one sub page of the `shop` page - used in the example above - would be possible to handle such urls.

**Comment by fritzmg** [comment]
I was thinking the same, but then you have the problem of having two pages in your site structure, both of which have valid, but different URLs.

* `https://example.org/shop/product-lorem-ipsum.html` and
* `https://example.org/product-detail/product-lorem-ipsum.html` 

both show the same content. Of course Isotope takes care of that by not writing the latter into the sitemap or showing this URL anywhere else - but still, I don't think this is an ideal solution.

**Comment by leofeyer** [comment]
I don't think there is a solution to this problem at all – except using separate pages for the news list and reader modules.

@frontendschlampe Maybe you can set up some kind of rewrite rule that internally rewrites requests with an item `en/news/foo.html` to `en/news-detail/foo.html`?

**Comment by fritzmg** [comment]
> @frontendschlampe Maybe you can set up some kind of rewrite rule that internally rewrites requests with an item `en/news/foo.html` to `en/news-detail/foo.html`?

The output of the URLs would also need to be changed.

Imho this is something worth looking into. May be we can work out a concept for stuff like this.

**Comment by frontendschlampe** [comment]
The rewrite rule is not the problem. The problem is the additional URL-fragment ```news-detail``` and it's there even if there's a rewrite rule! ;-)

We should have this thing in mind for the future!

**Comment by leofeyer** [comment]
> The problem is the additional URL-fragment `news-detail` and it's there

Of course you have to adjust the target page of the new archive to point to the `news` page.

>> The news list page includes in main area an intro headline, a text and the news list module for all news archives. In right column is a news archive navigation and a contact teaser.
>>
>> The detail page don't have any intro headline and text in main area, but the news detail module. In right column there's no news archive navigation, but a latest news list for the same news archive of the news in details and a contact teaser of the editor of the news.
>
> Imho this is something worth looking into.

I think that this is a very special case that the majority of users does not have. But let's see what others say.

**Comment by frontendschlampe** [comment]
> Of course you have to adjust the target page of the new archive to point to the news page.

You understand me wrong: I mean that a redirect doesn't change anything. It's just add another possibility to reach one and the same page. A redirect could have negative effects for the website, too: per news I have two URL

> I think that this is a very special case that the majority of users does not have. But let's see what others say.

Then I have special clients because nearly every client wants to have it. ;-)

**Comment by fritzmg** [comment]
Imho it's not a special case at all. The URL structure
```
example.org/news
example.org/news/news-item
```
is pretty common, I think.

**Comment by leofeyer** [comment]
That was not what I referred to (see what I have quoted above).

**Comment by fritzmg** [comment]
Ah, you mean the different layout - I still agree with @frontendschlampe there, we usually do not have the same content/layout for the list and the detail page when it comes to news entries.

**Comment by leofeyer** [comment]
As discussed in Mumble on December 20th, this can be solved with CMF routing, which we are currently working on in #95.

**Comment by leofeyer** [comment]
Implemented in #3843.
